### PR TITLE
IR refactor: simplify symbol table validation of identifier accesses

### PR DESCRIPTION
**Pull request by grjte**
This is the second in a series of PRs to address #150 . This one begins to simplify the symbol table, and especially the handling of references to identifiers. It is built on top of #179 . I haven't updated documentation yet because I wanted to see what people think of the approach first

### CODE REVIEW

**Review by bobbinth — APPROVED** [review]
Thank you! I like this structure - it looks much cleaner.

One potential suggestion I have is maybe to rename `IdentifierType` into `Symbol`. It seems to me that data contained in this enum is more than just type information. This also will clear up some confusing naming patters. For example, in:
```Rust
let symbol_type = symbol_table.get_type(symbol_name)?;
```
`get_type()` returns `IdentifierType` but we name it `symbol_type`. I'd probably have it as:
```Rust
let symbol = symbol_table.get_symbol(symbol_name)?;
```
And then `get_symbol()` would return `Symbol` type.

This could also make naming easier throughout. For example, `ValidateIdentifierAccess` trait could be named `SymbolAccessValidator`.

It is entirely possible that this naming makes sense just to me - so, I don't have a strong preference here.

**Comment by grjte** [comment]
> Thank you! I like this structure - it looks much cleaner.
> 
> One potential suggestion I have is maybe to rename `IdentifierType` into `Symbol`. It seems to me that data contained in this enum is more than just type information. This also will clear up some confusing naming patters. For example, in:
> 
> ```rust
> let symbol_type = symbol_table.get_type(symbol_name)?;
> ```
> 
> `get_type()` returns `IdentifierType` but we name it `symbol_type`. I'd probably have it as:
> 
> ```rust
> let symbol = symbol_table.get_symbol(symbol_name)?;
> ```
> 
> And then `get_symbol()` would return `Symbol` type.
> 
> This could also make naming easier throughout. For example, `ValidateIdentifierAccess` trait could be named `SymbolAccessValidator`.
> 
> It is entirely possible that this naming makes sense just to me - so, I don't have a strong preference here.

Yes, this is in a future PR. Breaking this stuff up into review-sized chunks has been a challenge in itself

File: ir/src/symbol_table/mod.rs
Comment (tohrnii) [inline ir/src/symbol_table/mod.rs:264]: nit: Any reason to panic here instead of returning an error?

**Review by tohrnii — APPROVED** [review]
Thank you @grjte, this is pretty neat. I just left a couple of non-blocking nits inline.

File: ir/src/symbol_table/access_validation.rs
Comment (grjte) [reply ir/src/symbol_table/access_validation.rs:12]: I did that first, but then I was thinking about updating the validation of the trace and random values accesses in as well, and those would be handled slightly differently. In essence, they would be `ValidateAccess` and these would be `ValidateIdentifierAccess`. I decided to hold off on that change, but kept the naming here for clarity

File: ir/src/symbol_table/mod.rs
Comment (grjte) [reply ir/src/symbol_table/mod.rs:264]: Yes - the reason I did this was that I want to move all the errors to the validation. This error is already handled in the `validate` call above, so the else case here should never happen. If it does happen, then something has gone badly wrong with the validation. I could put this into an `unreachable` instead for clarity if you think that's better

File: ir/src/symbol_table/mod.rs
Comment (grjte) [reply ir/src/symbol_table/mod.rs:93]: I think I was looking at different types of mutators and therefore trying to be more specific for better organization and clarity. However, it looks like that's no longer applicable. I can change this to `MUTATORS` and if I update in the future do `MUTATORS: SYMBOL INSERTIONS` instead

**Comment by grjte** [comment]
I'm planning to wait to merge this until we've merged #179

File: ir/src/symbol_table/mod.rs
Comment (tohrnii) [reply ir/src/symbol_table/mod.rs]: Thanks, I thought that could have been the reasoning. I don't have a strong opinion but to me `unreachable` does seem clearer for these cases but I'll defer to you.

**Review by Overcastan — APPROVED** [review]
Looks great, thank you @grjte!

**Pull request merged**
